### PR TITLE
specs, add specs for azure-core LRO

### PR DESCRIPTION
**Pull request by weidongxu-microsoft**
Guide https://github.com/microsoft/api-guidelines/blob/vNext/azure/ConsiderationsForServiceDesign.md#long-running-operations

Test does not include `LongRunningResourceCreateOrUpdate`, as guide does not recommend LRO on PATCH.

[Playground of this test](https://cadlplayground.z22.web.core.windows.net/cadl-azure/?c=aW1wb3J0ICJAYXp1cmUtdG9vbHMvY2FkbC3GEWNvcmUiOwrJJ8UbbGFuZy9yZXN0IjsKCnVzaW5nIEHELi5Db3JlO8cSZ2xvYmFsLsoZLlRyYWl0c8ggQ2FkbC5SZXN0zRFIdHRwOwoKI3N1cHByZXNz%2FgCaL2NhxDstc3R5bGUiICJGb3Igc3BlYyIKQGRvYygiSWxsdXN0cmF0ZXMgYm9kaWVzIHRlbXBsYXRlZCB3aXRo5gC4IOQAn8YQbG9uZy1ydW5u5ACSb3BlcmF0aW9uIikKQHNlcnZpY2UoewogIHRpdGxlOiAiTG9uZ1LGLSIsCiAgdmVyc2lvbjogIjEuMC4wIiwKfSkKbmFtZXNwYWNlxm0uTHJv5wEpCmludGVyZsQbUmVzb3VyY2VPyHNzCiAgZXh0ZW5kc%2FMBS9IvPE5vQ29uZGnEDWFsUmVxdWVzdHMgJiBOb1JlcGVhdGFibGXNF0NsaWVudMcTSWQ%2BIHt9CgpAcsdaKCJ1c2Vyc%2BQA%2BuUBS0RldGFpbHMgYWJvdXQgYSDEHi4iKQptb2RlbCBVc2VyIOQBGkBrZXnEB8U0VGhlIOQA%2FiBvZskwICBAdmlzaWJpbGl0eSgicmVhZMUWxCg6IHN0cmluZzsKzUNkZXNjcmlw5ADayErFM8sYyjrkAMTJOnBhcmFtZXRlcnMgZm9yIGV45ALG5AHa9ADARcUeUMQzc%2BYAzHF1ZXLuAM5mb3JtYXTkAIZ05ACZYXRhxEQgxxj3AIbGd2VkxXTJNcZ5xnVlZOoBQfsBOvsBJMlkVVJJxi3oAclVcmnzAJhDcmXlAxZvciByZXBhbGNlcyBh5QEH5QLsdW1tYXJ5KCJBZGTEGOUAuc4rxBMncyBmaWVsZHPEckBwb2xsaW5n6QKTKPICpS5HZXTRFlN0YXR1czzEfCwgbmV2ZXI%2BKQpvcCBj5QCjT3JSZXBs5AMjaXPzAyYu6wODyB7mANzJOcVaPjvoAPpEZWxl5AD68wDuyhvoAXT%2FANX%2FANXzANVkxWP%2FAMzrAMzmAJPvAMPmAjX1AMPKG%2F8Aw%2F8Aw%2BoAw%2BwCouYAxOYCbP8AxOsAxEFjxBvnAQzwA2os7QMDPjsK)

SDK is trying to solve the naming of the response of `LongRunningResourceDelete` and `LongRunningResourceAction`.

Currently SDK likely not able to properly name the templated model. E.g. [incorrect in Java](https://github.com/Azure/autorest.java/blob/9c31cb21ed839ddd9d5fd3e584cdee0dc92e6fc1/cadl-tests/src/main/java/com/cadl/longrunning/models/ResourceOperationStatusResourceExportedResourceError.java)

SDK is checking if we need some model in language core.

# Cadl Ranch Contribution Checklist:

- [x] I have written a [scenario spec](../docs/writing-scenario-spec.md)
- [x] I have **meaningful** `@scenario` names. Someone can look at the list of scenarios and understand what I'm covering.
- [ ] I have written a [mock API](../docs/writing-mock-apis.md)
- [ ] I have used `@scenarioDoc`s for extra scenario description and to tell people **how to pass** my mock api check.

### CODE REVIEW

File: packages/cadl-ranch-specs/http/lro/lro-core/main.cadl [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  "id":"operation1",
+  "status":"Succeeded",
+  "result":{
+    "name":"Madge",
+    "resourceUri":"/users/1"
+  }
+}
+```
+""")
+@pollingOperation(ResourceOperations.GetResourceOperationStatus<User, ExportedUser>)
```
Comment (weidongxu-microsoft) [inline packages/cadl-ranch-specs/http/lro/lro-core/main.cadl:130]: status monitor for POST (Action) be `id + status + error + result`

File: packages/cadl-ranch-specs/http/lro/lro-core/main.cadl [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  "status":"Succeeded"
+}
+
+{
+  "name": "Madge"
+  "description": "User Madge"
+}
+```
+""")
+@pollingOperation(ResourceOperations.GetResourceOperationStatus<User, never>)
```
Comment (weidongxu-microsoft) [inline packages/cadl-ranch-specs/http/lro/lro-core/main.cadl:79]: status monitor for PUT and DELETE be `id + status + error + result`

File: packages/cadl-ranch-specs/http/lro/lro-core/main.cadl [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+Expected path parameter: name=madge
+
+Expected input body:
+```json
+{
+  "role": "contributor"
+}
+```
+
+Expected response body:
```
Comment (mikekistler) [inline packages/cadl-ranch-specs/http/lro/lro-core/main.cadl:61]: The response of an LRO PUT should the resource -- a User in this case.

File: packages/cadl-ranch-specs/http/lro/lro-core/main.cadl [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+```json
+{
+  "id":"operation1",
+  "status":"InProgress"
+}
+
+{
+  "id":"operation1",
+  "status":"Succeeded"
+}
```
Comment (mikekistler) [inline packages/cadl-ranch-specs/http/lro/lro-core/main.cadl:98]: The second response is from the pollingOperation, right?

File: packages/cadl-ranch-specs/http/lro/lro-core/main.cadl (hunk 3 above)
Comment (weidongxu-microsoft) [reply packages/cadl-ranch-specs/http/lro/lro-core/main.cadl:61]: Yes, the final result is the User resource.

The first 2 in expected responses describes the polling operation. For Java, these poll result also visible to user, in [SyncPoller/PollerFlux](https://azure.github.io/azure-sdk/java_introduction.html#java-lro-poller-class) (the `T` type)

File: packages/cadl-ranch-specs/http/lro/lro-core/main.cadl (hunk 4 above)
Comment (weidongxu-microsoft) [reply packages/cadl-ranch-specs/http/lro/lro-core/main.cadl:98]: Yes. Both (or at least the DELETE response) should also should includes a `operation-location` header (I didn't write here).

File: packages/cadl-ranch-specs/http/lro/lro-core/main.cadl (hunk 1 above)
Comment (weidongxu-microsoft) [reply packages/cadl-ranch-specs/http/lro/lro-core/main.cadl:130]: Ok. We won't include this case in test. Here we will only do success case.

File: packages/cadl-ranch-specs/http/lro/lro-core/main.cadl [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  "status":"InProgress"
+}
+
+{
+  "id":"operation1",
+  "status":"Succeeded",
+  "result":{
+    "name":"madge",
+    "resourceUri":"/users/1"
+  }
```
[the remark this answers is not in the document]
Comment (weidongxu-microsoft) [reply packages/cadl-ranch-specs/http/lro/lro-core/main.cadl:126]: Yes, it comes back from polling.

[the remark this answers is not in the document]
Comment (weidongxu-microsoft) [reply packages/cadl-ranch-specs/http/lro/lro-core/main.cadl:126]: And indeed this is a great question on possible SDK design.

E.g. as Java uses patten as `SyncPoller<PollResponseType, FinalResultType>` as response of the LRO API, we'd like to discuss whether we'd generate API
```java
SyncPoller<{poll-response-that-to-be-decided}, ExportedUser> export(...)
```
which extract the `ExportedUser` object from the last success `result` property and put it as the final result.

Or merely
```java
SyncPoller<{poll-response-class-that-to-be-decided}, Void> export(...)
```
that reflexing that everything is via pollingOperation.

File: packages/cadl-ranch-specs/http/lro/lro-core/main.cadl (hunk 3 above)
Comment (weidongxu-microsoft) [reply packages/cadl-ranch-specs/http/lro/lro-core/main.cadl:64]: I've enhance the `@scenarioDoc` to spell out details on every request/response.

File: packages/cadl-ranch-specs/http/lro/lro-core/main.cadl [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+Expected URL: {endpoint}/users/madge
+
+Expected status code: 200
+Expected response body:
+```json
+{
+  "name": "madge"
+  "role": "contributor"
+}
+```
```
Comment (weidongxu-microsoft) [inline packages/cadl-ranch-specs/http/lro/lro-core/main.cadl:106]: In old "PUT to create" or "POST to create" scenario (the "operation resource" pattern), the last "status=Succeeded" response of status monitor would contains a "resourceLocation" property to point to the URL of the resource created.

This won't be there in this new design.

In Guideline, the last GET on resource URL (`{endpoint}/users/madge`) is not required. But I am not sure if SDK would still want to do this, to get the final result (in case any property of the resource changes from the first 201 from PUT and this 200 from GET).

File: packages/cadl-ranch-specs/http/lro/lro-core/main.cadl [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+Expected request body:
+```json
+{
+  "role": "contributor"
+}
+```
+
+Expected status code: 201
+Expected response header: operation-location={endpoint}/users/madge/operations/operation1
```
Comment (weidongxu-microsoft) [inline packages/cadl-ranch-specs/http/lro/lro-core/main.cadl:63]: The status monitor URL from current Azure.Core is not exactly same as from the Guideline (note the format of the status monitor URL is not a requirement from Guideline).

This URL should be opaque to SDK, so for now I adopt the one from `Azure.Core.ResourceOperations.GetResourceOperationStatus<User, ..>`

**Comment by weidongxu-microsoft** [comment]
It may relate to https://github.com/Azure/typespec-azure/issues/2675 on the detail of the runtime processing on LRO PUT.

**Comment by iscai-msft** [comment]
@weidongxu-microsoft looks good! if you have time before this pr gets merged can you add a test for LRORpcOperation? It's not currently in core (afaik), but health decision is using it and it plans to go into core. I think for now we can vendor the code like they're doing (code [here](https://github.com/Azure/azure-rest-api-specs/pull/22990/files#diff-c442a8f83e02064e254d7dad175689aeb4da1fc416ec7f90559daea773b50871R29)). Would be great to test this for health decision since it's a high priority service

**Comment by weidongxu-microsoft** [comment]
> @weidongxu-microsoft looks good! if you have time before this pr gets merged can you add a test for LRORpcOperation? It's not currently in core (afaik), but health decision is using it and it plans to go into core. I think for now we can vendor the code like they're doing (code [here](https://github.com/Azure/azure-rest-api-specs/pull/22990/files#diff-c442a8f83e02064e254d7dad175689aeb4da1fc416ec7f90559daea773b50871R29)). Would be great to test this for health decision since it's a high priority service.
> 
> One other good test to add is where the result of the polling operation is different than the final logical result we want to return. Thanks so much!

Sure, I will check the LRORpcOperation, and add it accordingly.

In current, LongRunningResourceCreateOrReplace is the case that poll response be different from final response.

The LongRunningResourceAction likely need certain sync with architects, as the "logical result" is the object in "result" property of last success poll response (this is new).
At present with legacy code, SDK likely will be satisfied use the last poll response as final response.

---

We can sure to wait a few days for this to merge. (azure-core in Java with legacy "operation-location" does not work for this PUT, I am still working on it, hence myself not really completely tested the PUT scenario).

**Comment by weidongxu-microsoft** [comment]
@iscai-msft

I've put the LongRunningRpcOperation PR at https://github.com/Azure/cadl-ranch/pull/242

`pnpm generate-scenarios-summary` might not work correctly after rename to tsp (I guess it reads for .cadl). Haven't got chance to check.

I've checked current LroMetadata output, seems good on `logicalResult`, but maybe not good on `statusMonitorStep.responseModel`
Here is
```yaml
finalStep: undefined
logicalResult: Model JobResult
statusMonitorStep:
  responseModel: Model OperationStatus
```

For it to get
```yaml
statusMonitorStep:
  responseModel: Model JobResult
```

I need to use a customized `LongRunningStatusLocation`
```ts
  @doc("Metadata for long running operation status monitor locations")
  model LongRunningStatusLocation<TResponse extends object> {
    @pollingLocation
    @doc("The location for monitoring the operation state.")
    @TypeSpec.Http.header("Operation-Location")
    operationLocation: Rest.ResourceLocation<TResponse>;
  }

  @doc("Long running RPC operation template")
  op LongRunningRpcOperation<
      TParams extends object,
      TResponse extends object,
      Traits extends object = {}
    > is RpcOperation<
      TParams & Traits,
      Foundations.AcceptedResponse<Traits &
                                  LongRunningStatusLocation<TResponse> &
                                  Foundations.RetryAfterHeader &
                                  RepeatabilityResponseHeaders> | TResponse,
      Traits
    >;
```

I think Mark's helper follows `ResourceLocation<>` template for the poll response.

**Comment by weidongxu-microsoft** [comment]
If I need to get the `LongRunningRpcOperation` return different poll response and final response, I will need to write this to get `LroMetadata` work

```ts
  @doc("Metadata for long running operation status monitor locations")
  model LongRunningStatusLocation<TPollResponse extends object> {
    @pollingLocation
    @doc("The location for monitoring the operation state.")
    @TypeSpec.Http.header("Operation-Location")
    operationLocation: Rest.ResourceLocation<TPollResponse>;
  }

  @doc("Long running RPC operation template")
  op LongRunningRpcOperation<
      TParams extends object,
      TResponse extends object,
      TPollResponse extends object = TResponse,
      Traits extends object = {}
    > is RpcOperation<
      TParams & Traits,
      Foundations.AcceptedResponse<Traits &
                                  LongRunningStatusLocation<TPollResponse> &
                                  Foundations.RetryAfterHeader &
                                  RepeatabilityResponseHeaders> | TResponse,
      Traits
    >;
```

**Comment by weidongxu-microsoft** [comment]
I've added a case of different poll / final location / result to the PR https://github.com/Azure/cadl-ranch/pull/242#discussion_r1147096385

However, my guess is that the final one to Azure.Core likely not be same as what I have here.

**Review by iscai-msft — APPROVED** [review]
